LEAN-3978 LW-FY25-Q2-S3: Table_element delete suggestion is not showi…

fwaisi opened the pull request:
This PR addresses the following issues related to cutting and pasting table content:

- Cutting part of a table and pasting it: The paste functionality was not working correctly, and I have fixed it.

- Cutting the whole table: Previously, a delete suggestion was shown for all table cells text content. I have corrected this to show only a delete suggestion for the table_element.

- Cutting part of the article content containing a table: The table content was being pasted as text. This has been resolved so the table is pasted correctly.

IslamJMomani reviewed (commented):
The code needs formatting; I see many of `;` here.

IslamJMomani commented on src/lib/copy.ts:
```diff
@@ -0,0 +1,71 @@
+      let tableStart: number | null = null;
+      let tableNode: Node | null = null;
+  
+      // Find the table_element node that contains the copied content
```
could this be simplified by using `findParentNodeOfType`?

fwaisi replied:
done.

IslamJMomani commented on src/lib/copy.ts:
```diff
@@ -0,0 +1,71 @@
+      }
+    }
+  
+    // handle case when part of the copied content is a table
```
Could `findChildrenByType` be useful in that case?

fwaisi replied:
> Can you please elaborate on this case more?

this logic handle when the user cut multiple nodes that contains table node. e.g. paragraph table

fwaisi replied:
> Could `findChildrenByType` be useful in that case?

I don't think so, because here I am looking for the children count not a specific node type.

asouqi reviewed (changes requested):
I checked the clipboard and paste, it parse HTML to the correct slice. our issue here with the track-changes as the result of that paste is a ReplaceAround step that cuts text selection with closed content. if we turn off the track plugin it's going to work.

The fix in TC will be from the [slice](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/blob/master/src/steps/trackReplaceAroundStep.ts#L119) of `insert-slice`, in our case content of the slice is not compatible with the location of text selection so it needs to be wrapped with matched content(similar to list-schema commands). we need to do something like this:
```
if (tr.getMeta('uiEvent') === 'paste') {
        if (slice.openStart === 1) {
            fragment = fragment.addToStart(oldState.selection.$anchor.node().type.create())
        }

        if (gap.size > 0) {
            fragment = fragment.append(setFragmentAsNodeSplit(newTr.doc.resolve(step.from), newTr, Fragment.from(oldState.selection.$anchor.node().type.create(undefined, gap.content)), attrs))
        }

        steps.push({
            type: 'insert-slice',
            from: gapFrom,
            to: gapTo + 1,
            slice: new Slice(fragment, slice.openStart, slice.openEnd),
            sliceWasSplit,
        });
        return steps;
    }
``` 
in [trackReplaceAroundStep](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/blob/master/src/steps/trackReplaceAroundStep.ts#L99)


> Cutting the whole table: Previously, a delete suggestion was shown for all table cells text content. I have corrected this to show only a delete suggestion for the table_element.

where is the fix of the table selection cut! I can't see that in this PR. can you please point to me where is the files of that fix

fwaisi commented on src/lib/copy.ts:
```diff
@@ -0,0 +1,71 @@
+        if (node.type === schema.nodes.table_element) {
+          const nodeEnd = pos + (node.nodeSize || 0);
+          if (pos <= from && nodeEnd >= to) {
+            const tableElement = getTableElementNode(state);
```
It appears to be a repeated task, and I'm not sure why I included it initially.

fwaisi commented:
> I checked the clipboard and paste, it parse HTML to the correct slice. our issue here with the track-changes as the result of that paste is a ReplaceAround step that cuts text selection with closed content. if we turn off the track plugin it's going to work.
> 
> The fix in TC will be from the [slice](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/blob/master/src/steps/trackReplaceAroundStep.ts#L119) of `insert-slice`, in our case content of the slice is not compatible with the location of text selection so it needs to be wrapped with matched content(similar to list-schema commands). we need to do something like this:
> 
> ```
> if (tr.getMeta('uiEvent') === 'paste') {
>         if (slice.openStart === 1) {
>             fragment = fragment.addToStart(oldState.selection.$anchor.node().type.create())
>         }
> 
>         if (gap.size > 0) {
>             fragment = fragment.append(setFragmentAsNodeSplit(newTr.doc.resolve(step.from), newTr, Fragment.from(oldState.selection.$anchor.node().type.create(undefined, gap.content)), attrs))
>         }
> 
>         steps.push({
>             type: 'insert-slice',
>             from: gapFrom,
>             to: gapTo + 1,
>             slice: new Slice(fragment, slice.openStart, slice.openEnd),
>             sliceWasSplit,
>         });
>         return steps;
>     }
> ```
> 
> in [trackReplaceAroundStep](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/blob/master/src/steps/trackReplaceAroundStep.ts#L99)
> 
> > Cutting the whole table: Previously, a delete suggestion was shown for all table cells text content. I have corrected this to show only a delete suggestion for the table_element.
> 
> where is the fix of the table selection cut! I can't see that in this PR. can you please point to me where is the files of that fix

fwaisi commented:
> I checked the clipboard and paste, it parse HTML to the correct slice. our issue here with the track-changes as the result of that paste is a ReplaceAround step that cuts text selection with closed content. if we turn off the track plugin it's going to work.
> 
> The fix in TC will be from the [slice](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/blob/master/src/steps/trackReplaceAroundStep.ts#L119) of `insert-slice`, in our case content of the slice is not compatible with the location of text selection so it needs to be wrapped with matched content(similar to list-schema commands). we need to do something like this:
> 
> ```
> if (tr.getMeta('uiEvent') === 'paste') {
>         if (slice.openStart === 1) {
>             fragment = fragment.addToStart(oldState.selection.$anchor.node().type.create())
>         }
> 
>         if (gap.size > 0) {
>             fragment = fragment.append(setFragmentAsNodeSplit(newTr.doc.resolve(step.from), newTr, Fragment.from(oldState.selection.$anchor.node().type.create(undefined, gap.content)), attrs))
>         }
> 
>         steps.push({
>             type: 'insert-slice',
>             from: gapFrom,
>             to: gapTo + 1,
>             slice: new Slice(fragment, slice.openStart, slice.openEnd),
>             sliceWasSplit,
>         });
>         return steps;
>     }
> ```
> 
> in [trackReplaceAroundStep](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/blob/master/src/steps/trackReplaceAroundStep.ts#L99)
> 
> >

I don’t think we need to update anything in the track-changes plugin because the issue primarily lies in how ProseMirror processes cut and copy operations, which seems to be expected behavior. However, we are trying to implement a special case for the table element node.

Case Breakdown:

- Cutting the Whole Table Node: When selecting the entire table node and performing a cut, ProseMirror generates suggestions for each table_cell text content. This is the default behavior of the ProseMirror table plugin. In this scenario, we want to cut the entire table_element node, including related components like footnotes and figcaptions. To achieve this, I intercepted the copy functionality and updated the slice to include the target table_element instead of just the table node. This change ensures that the correct suggestions are generated when cutting the table_element.

You can find the relevant code responsible for this in the following link: [copy.ts](https://github.com/Atypon-OpenSource/manuscripts-body-editor/blob/1f524f663a3cf59a6f9c44723f4dec73c0da5030/src/lib/copy.ts#L26).

- Cutting a Highlighted Paragraph and the Table Node: In this case, when a paragraph is highlighted alongside the table node and a cut operation is performed, pasting the content results in malformed table content. The issue arises because the listing_element node is not cut along with the table node, which leads to an incorrect structure for the table_element. Consequently, when pasted, the table is treated as plain text rather than retaining its original formatting.

asouqi commented:
> I don’t think we need to update anything in the track-changes plugin because the issue primarily lies in how ProseMirror processes cut and copy operations, which seems to be expected behavior. However, we are trying to implement a special case for the table element node.

What is the special case for table? for copying and pasting a table from editor ProseMirror work with no issue at all, it's going to be parsed to the right slice, and if we disable the track-changes-plugin ProseMirror add that slice without a problem! 

> When selecting the entire table node and performing a cut, ProseMirror generates suggestions for each table_cell text content. This is the default behavior of the ProseMirror table plugin. In this scenario, we want to cut the entire table_element node, including related components like footnotes and figcaptions. To achieve this, I intercepted the copy functionality and updated the slice to include the target table_element instead of just the table node. This change ensures that the correct suggestions are generated when cutting the table_element.

The fix should be from the selection, so if we select all table cells selection should point to `table_element` node. we should not handle that from the `transformCopied`

IslamJMomani commented:
@fwaisi, I have some concerns around this PR that need to be discussed. As far as I can see, @asouqi also has some points, so I think it’s good to arrange a call, maybe tomorrow (on Thursday), for the three of us to discuss it.

asouqi commented:
> I believe implementing this fix within the track-changes repo would be most appropriate, as it appears to be the root cause.

yes, and the fix of **just paste with ReplaceAroundStep** will be in track-changes. cut fix will be in body-editor so we can [intercept](https://github.com/Atypon-OpenSource/manuscripts-body-editor/blob/master/src/plugins/tables-cursor-fix.ts#L24) if selection about the table by using [isInTable](https://github.com/ProseMirror/prosemirror-tables/blob/d4487a021054e1cc48707140376d01e1db802043/src/util.ts#L51) and move selection to table element node

IslamJMomani commented:
> * Updating the selection when an entire table node is selected to include related elements (figure captions, footnotes, etc.)
> * Creating a table node from selected table cells and rows when performing a paste operation on cut portions of the table
> Could you please advise whether these implementations should be handled in track-changes or body-editor?

@fwaisi, In my opinion, these customizations could be done in the body-editor.

> I will address the table cut-and-paste issue that occurs when the selection contains both section and table nodes. I believe implementing this fix within the track-changes repo would be most appropriate, as it appears to be the root cause.

In this case, are you referring to selecting more content (maybe paragraphs) before and after a table element? If so, I see it's already working.

We can discuss this further on the call.

mbartenev-atypon commented:
> > * Updating the selection when an entire table node is selected to include related elements (figure captions, footnotes, etc.)
> > * Creating a table node from selected table cells and rows when performing a paste operation on cut portions of the table
> >   Could you please advise whether these implementations should be handled in track-changes or body-editor?
> 
> @fwaisi, In my opinion, these customizations could be done in the body-editor.
> 
> > I will address the table cut-and-paste issue that occurs when the selection contains both section and table nodes. I believe implementing this fix within the track-changes repo would be most appropriate, as it appears to be the root cause.
> 
> In this case, are you referring to selecting more content (maybe paragraphs) before and after a table element? If so, I see it's already working.
> 
> We can discuss this further on the call.

I agree - it should be in body editor

mbartenev-atypon commented:
@fwaisi also please note that we plan to address cases, when mutiple changes need to be represented as a single, in composite changes epic. There are several cases where we need to address it would be preferable to approach it as a single problem.

fwaisi commented:
@asouqi, @IslamJMomani, I have removed the logic related to handle multiple nodes cut, and moved the solution to the [track-changes-plugin](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/pull/36).

@asouqi, using isInTable, it won't work properly since it will return true for both case cutting part of the table or cutting the whole table.

asouqi commented:
> using isInTable, it won't work properly since it will return true for both case cutting part of the table or cutting the whole table.

try this in the plugin:
```
const isTableElementShapeSelected = (state: EditorState) => {
  if (state.selection instanceof CellSelection) {
    const rect = selectedRect(state)
    return rect.bottom === rect.map.height && rect.right === rect.map.width
  }

  return false
}
```

This will be [in](https://github.com/Atypon-OpenSource/manuscripts-body-editor/blob/master/src/plugins/tables-cursor-fix.ts#L24) 
```
const cutTr = transactions.find((tr) => tr.getMeta('uiEvent') && tr.getMeta('uiEvent') === 'cut');

if (cutTr && isTableElementShapeSelected(oldState)) {
   const $table = findParentNodeOfTypeClosestToPos(newState.selection.$from, schema.nodes.table_element)
   if ($table) {
      const newTr = newState.tr;
      newTr.delete($table.pos, $table.pos + $table.node.nodeSize)
      return newTr;
      }
  }
```

> Cutting the whole table: Previously, a delete suggestion was shown for all table cells text content. I have corrected this to show only a delete suggestion for the table_element.

@fwaisi Panos confirmed that or is it from QA? as I can see in other editors like google doc cut will remove just text

fwaisi commented:
> > using isInTable, it won't work properly since it will return true for both case cutting part of the table or cutting the whole table.
> 
> try this in the plugin:
> 
> ```
> const isTableElementShapeSelected = (state: EditorState) => {
>   if (state.selection instanceof CellSelection) {
>     const rect = selectedRect(state)
>     return rect.bottom === rect.map.height && rect.right === rect.map.width
>   }
> 
>   return false
> }
> ```
> 
> This will be [in](https://github.com/Atypon-OpenSource/manuscripts-body-editor/blob/master/src/plugins/tables-cursor-fix.ts#L24)
> 
> ```
> const cutTr = transactions.find((tr) => tr.getMeta('uiEvent') && tr.getMeta('uiEvent') === 'cut');
> 
> if (cutTr && isTableElementShapeSelected(oldState)) {
>    const $table = findParentNodeOfTypeClosestToPos(newState.selection.$from, schema.nodes.table_element)
>    if ($table) {
>       const newTr = newState.tr;
>       newTr.delete($table.pos, $table.pos + $table.node.nodeSize)
>       return newTr;
>       }
>   }
> ```

@asouqi, I tried using uiMeta within copy.ts, but it's coming up as undefined. Also, the plugin doesn't seem to be triggered when performing a cut operation. Since the current check is achieving the intended result, I don't see the need to reimplement it.


> @fwaisi Panos confirmed that or is it from QA? as I can see in other editors like google doc cut will remove just text
@asouqi, Please check the following comment: https://jira.atypon.com/browse/LEAN-3978?focusedCommentId=6989362#comment-6989362

asouqi commented:
> I tried using uiMeta within copy.ts, but it's coming up as undefined. Also, the plugin doesn't seem to be triggered when performing a cut operation. Since the current check is achieving the intended result, I don't see the need to reimplement it.

try this patch on master without this PR changes
[cell_selection.zip](https://github.com/user-attachments/files/17492702/cell_selection.zip)

asouqi commented on src/lib/copy.ts:
```diff
@@ -0,0 +1,58 @@
+        )
+      }
+    }
+    if (tableNode) {
+      return new Slice(Fragment.from(tableNode), 0, 0)
+    }
```
we can do it like this:
```
     if (slice.openStart !== slice.openEnd && (slice.content.firstChild?.type.isBlock || slice.content.lastChild?.type.isBlock)) {
                const cutDepth = Math.min(slice.openStart, slice.openEnd)
                return new Slice(slice.content, cutDepth, cutDepth)
            }
```
to balance slice sides for block elements as that will help track-changes plugin to deal with complex steps. and that should be removed after we fix different slice-side issue from TC.

asouqi commented on src/lib/paste.ts:
```diff
@@ -85,6 +87,23 @@ export const handlePaste = (
     tr.selection
   )
+
+  if (
+    slice.content.firstChild &&
+    (slice.content.firstChild.type == schema.nodes.table_row ||
+      slice.content.firstChild.type == schema.nodes.table_cell)
+  ) {
+    const newTable = createTableFromSlice(slice, view.state.schema)
+    if (newTable) {
+      const tr = view.state.tr.insert(
+        view.state.selection.$anchor.pos,
+        newTable
+      )
+      view.dispatch(tr)
+      return true
+    }
+  }
```
Issue of copying cells or rows related to the structure of [wrapMap](https://github.com/ProseMirror/prosemirror-view/blob/0c3e5d35baf8af161dd2ed7b2e566c75362abfbb/src/clipboard.ts#L197-L199) as it assumes will be `table -> tbody` but in our case that will be `figure -> table -> tbody` figure which is [tableElement](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/master/src/schema/nodes/table_element.ts#L41C1-L48C9) node in DOM, so if you change that code locally to add `figure` at first in each array issue will be fixed.

We can fix that from our side by using `transformPastedHTML` in [EditorView](https://github.com/Atypon-OpenSource/manuscripts-body-editor/blob/6d24b2031be40663f75b20d2f03d089ca3ac5582/src/configs/ManuscriptsEditor.ts#L82)
```
transformPastedHTML: (html) => {
      // add figure which is table_element node in DOM
      if (html.includes('table')) {
        const doc = new DOMParser().parseFromString(html, 'text/html');
        doc.querySelectorAll('table').forEach((value) => {
          if (value.parentElement?.tagName !== 'figure') {
            value.parentElement?.insertBefore(document.createElement('figure'), value)
          }
        })
        return doc.body.innerHTML
      }
      return html
    }
```

mbartenev-atypon commented:
@fwaisi , do we need this PR still?

fwaisi commented:
@mbartenev-atypon, no, closing.
